File: .github/scripts/test-profile-contract-model.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "tmpdir"
require_relative "lib/profile_contracts"

failures = []

check = lambda do |name, &block|
  block.call
rescue StandardError => error
  failures << "#{name}: #{error.class}: #{error.message}"
end

assert = lambda do |condition, message|
  raise message unless condition
end

check.call("parses one normalized profile declaration") do
  Dir.mktmpdir("profile-contract-model-test") do |directory|
    path = File.join(directory, "SKILL.md")
    File.write(path, "- Selected profiles: `packaged-cli, mcp-enabled`\n")
    selection = ProfileContracts::ProfileSelection.load(path)
    assert.call(selection.profiles == %w[packaged-cli mcp-enabled], selection.profiles.inspect)
    assert.call(selection.selected?("packaged-cli"), "packaged-cli was not selected")
    assert.call(!selection.template_scaffold?, "concrete selection reported as scaffold")
  end
end

check.call("rejects duplicate profile declarations") do
  Dir.mktmpdir("profile-contract-model-test") do |directory|
    path = File.join(directory, "SKILL.md")
    File.write(path, "Selected profiles: packaged-cli\nSelected profiles: mcp-enabled\n")
    begin
      ProfileContracts::ProfileSelection.load(path)
    rescue ProfileContracts::ParseError => error
      assert.call(error.message.include?("exactly one"), error.message)
      next
    end
    raise "duplicate declarations were accepted"
  end
end

check.call("rejects an empty profile declaration") do
  Dir.mktmpdir("profile-contract-model-test") do |directory|
    path = File.join(directory, "SKILL.md")
    File.write(path, "Selected profiles:\n")
    begin
      ProfileContracts::ProfileSelection.load(path)
    rescue ProfileContracts::ParseError => error
      assert.call(error.message.include?("at least one"), error.message)
      next
    end
    raise "empty profile declaration was accepted"
  end
end

check.call("extracts nested and peer Markdown sections") do
  document = ProfileContracts::MarkdownDocument.new(<<~MARKDOWN)
    ## Parent

    Parent value: retained

    ### Child

    Child value: retained

    ## Peer

    Peer value: excluded
  MARKDOWN
  parent = document.section("## Parent")
  child = document.section("### Child")
  assert.call(parent.include?("Child value: retained"), parent.inspect)
  assert.call(!parent.include?("Peer value: excluded"), parent.inspect)
  assert.call(child.include?("Child value: retained"), child.inspect)
end

check.call("normalizes scalar fields and table cells") do
  document = ProfileContracts::MarkdownDocument.new(<<~MARKDOWN)
    Mode selector: `--json`

    | Item | Value |
    |---|---|
    | TBD | `JSON` |
    | Failure | Validation \\| runtime failure |
  MARKDOWN
  assert.call(document.field("Mode selector") == "--json", document.field("Mode selector").inspect)
  assert.call(document.table_rows[1] == ["TBD", "JSON"], document.table_rows.inspect)
  assert.call(
    document.table_rows.last == ["Failure", "Validation \\| runtime failure"],
    document.table_rows.inspect
  )
  table_values = document.each_scalar.select { |entry| entry.kind == :table }.map(&:value)
  assert.call(table_values.include?("TBD"), table_values.inspect)
  assert.call(table_values.include?("Validation \\| runtime failure"), table_values.inspect)
end

check.call("applies shared unresolved and concrete value policy") do
  policy = ProfileContracts::ValuePolicy
  assert.call(policy.unresolved_scalar?("`TBD`"), "TBD was not unresolved")
  assert.call(policy.unresolved_scalar?("details forthcoming"), "forthcoming phrase was not unresolved")
  assert.call(!policy.unresolved_scalar?("documented behavior"), "concrete prose was unresolved")
  assert.call(policy.resolved?("NONE"), "NONE should be resolved")
  assert.call(!policy.concrete?("NONE"), "NONE should not be concrete")
  assert.call(policy.concrete?("successful completion"), "concrete value was rejected")
end

unless failures.empty?
  failures.each { |failure| warn failure }
  exit 1
end

puts "Shared profile contract model tests passed."

File: .github/scripts/test_profile_contract_model.py
#!/usr/bin/env python3
"""Parity tests for the shared Python profile-contract model."""

from __future__ import annotations

import sys
import tempfile
from pathlib import Path

from lib.profile_contracts import (
    MarkdownDocument,
    ParseError,
    ProfileSelection,
    SkillDocument,
    ValuePolicy,
)


def _assert(condition: bool, message: str) -> None:
    if not condition:
        raise AssertionError(message)


def run() -> int:
    failures: list[str] = []

    def check(name: str, operation) -> None:  # type: ignore[no-untyped-def]
        try:
            operation()
        except Exception as exc:  # noqa: BLE001 - aggregate all parity failures.
            failures.append(f"{name}: {type(exc).__name__}: {exc}")

    def parses_one_normalized_profile_declaration() -> None:
        with tempfile.TemporaryDirectory(prefix="profile-contract-model-test-") as directory:
            path = Path(directory) / "SKILL.md"
            path.write_text(
                "- Selected profiles: `packaged-cli, mcp-enabled`\n",
                encoding="utf-8",
            )
            selection = ProfileSelection.load(path)
            _assert(
                selection.profiles == ("packaged-cli", "mcp-enabled"),
                repr(selection.profiles),
            )
            _assert(
                selection.selected("packaged-cli"),
                "packaged-cli was not selected",
            )
            _assert(
                not selection.template_scaffold(),
                "concrete selection reported as scaffold",
            )

    def rejects_duplicate_profile_declarations() -> None:
        with tempfile.TemporaryDirectory(prefix="profile-contract-model-test-") as directory:
            path = Path(directory) / "SKILL.md"
            path.write_text(
                "Selected profiles: packaged-cli\n"
                "Selected profiles: mcp-enabled\n",
                encoding="utf-8",
            )
            try:
                ProfileSelection.load(path)
            except ParseError as exc:
                _assert("exactly one" in str(exc), str(exc))
                return
            raise AssertionError("duplicate declarations were accepted")

    def rejects_empty_profile_declaration() -> None:
        with tempfile.TemporaryDirectory(prefix="profile-contract-model-test-") as directory:
            path = Path(directory) / "SKILL.md"
            path.write_text("Selected profiles:\n", encoding="utf-8")
            try:
                ProfileSelection.load(path)
            except ParseError as exc:
                _assert("at least one" in str(exc), str(exc))
                return
            raise AssertionError("empty profile declaration was accepted")

    def extracts_nested_and_peer_markdown_sections() -> None:
        document = MarkdownDocument(
            """## Parent

Parent value: retained

### Child

Child value: retained

## Peer

Peer value: excluded
"""
        )
        parent = document.section("## Parent")
        child = document.section("### Child")
        _assert(parent is not None, "parent section was not found")
        _assert(child is not None, "child section was not found")
        _assert("Child value: retained" in parent, repr(parent))
        _assert("Peer value: excluded" not in parent, repr(parent))
        _assert("Child value: retained" in child, repr(child))

    def normalizes_scalar_fields_and_table_cells() -> None:
        document = MarkdownDocument(
            r"""Mode selector: `--json`

| Item | Value |
|---|---|
| TBD | `JSON` |
| Failure | Validation \| runtime failure |
"""
        )
        _assert(
            document.field("Mode selector") == "--json",
            repr(document.field("Mode selector")),
        )
        rows = document.table_rows()
        _assert(rows[1] == ["TBD", "JSON"], repr(rows))
        _assert(
            rows[-1] == ["Failure", r"Validation \| runtime failure"],
            repr(rows),
        )
        table_values = [
            entry.value for entry in document.each_scalar() if entry.kind == "table"
        ]
        _assert("TBD" in table_values, repr(table_values))
        _assert(r"Validation \| runtime failure" in table_values, repr(table_values))

    def applies_shared_unresolved_and_concrete_value_policy() -> None:
        _assert(ValuePolicy.unresolved_scalar("`TBD`"), "TBD was not unresolved")
        _assert(
            ValuePolicy.unresolved_scalar("details forthcoming"),
            "forthcoming phrase was not unresolved",
        )
        _assert(
            not ValuePolicy.unresolved_scalar("documented behavior"),
            "concrete prose was unresolved",
        )
        _assert(ValuePolicy.resolved("NONE"), "NONE should be resolved")
        _assert(not ValuePolicy.concrete("NONE"), "NONE should not be concrete")
        _assert(
            ValuePolicy.concrete("successful completion"),
            "concrete value was rejected",
        )

    def parses_safe_yaml_frontmatter() -> None:
        document = SkillDocument(
            """---
name: example-skill
description: Example
metadata:
  profiles:
    - instruction-only
---
# Example
""",
            path="SKILL.md",
        )
        _assert(document.metadata["name"] == "example-skill", repr(document.metadata))
        _assert(
            document.metadata["metadata"]["profiles"] == ["instruction-only"],
            repr(document.metadata),
        )

    def rejects_yaml_aliases() -> None:
        try:
            SkillDocument(
                """---
name: &identity example-skill
description: *identity
---
# Example
""",
                path="SKILL.md",
            )
        except ParseError as exc:
            _assert("aliases are not permitted" in str(exc), str(exc))
            return
        raise AssertionError("YAML alias was accepted")

    check(
        "parses one normalized profile declaration",
        parses_one_normalized_profile_declaration,
    )
    check(
        "rejects duplicate profile declarations",
        rejects_duplicate_profile_declarations,
    )
    check(
        "rejects an empty profile declaration",
        rejects_empty_profile_declaration,
    )
    check(
        "extracts nested and peer Markdown sections",
        extracts_nested_and_peer_markdown_sections,
    )
    check(
        "normalizes scalar fields and table cells",
        normalizes_scalar_fields_and_table_cells,
    )
    check(
        "applies shared unresolved and concrete value policy",
        applies_shared_unresolved_and_concrete_value_policy,
    )
    check("parses safe YAML frontmatter", parses_safe_yaml_frontmatter)
    check("rejects YAML aliases", rejects_yaml_aliases)

    if failures:
        for failure in failures:
            print(failure, file=sys.stderr)
        return 1

    print("Shared Python profile contract model tests passed (8 cases).")
    return 0


if __name__ == "__main__":
    raise SystemExit(run())
